server/test/services/question/types/IdQuestionDefinitionTest.java

```diff
@@ -0,0 +1,65 @@
+package services.question.types;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.OptionalInt;
+import junitparams.JUnitParamsRunner;
+import junitparams.Parameters;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import services.CiviFormError;
+import services.LocalizedStrings;
+import services.question.types.IdQuestionDefinition.IdValidationPredicates;
+
+@RunWith(JUnitParamsRunner.class)
+public class IdQuestionDefinitionTest {
+  @SuppressWarnings("unused") // Is used via reflection by the @Parameters annotation below
+  private static ImmutableList<Object[]> getValidationTestData() {
+    return ImmutableList.of(
+        new Object[] {OptionalInt.empty(), OptionalInt.empty(), Optional.<String>empty()},
+        new Object[] {OptionalInt.of(1), OptionalInt.empty(), Optional.<String>empty()},
+        new Object[] {OptionalInt.empty(), OptionalInt.of(1), Optional.<String>empty()},
+        new Object[] {OptionalInt.of(1), OptionalInt.of(2), Optional.<String>empty()},
+        new Object[] {OptionalInt.of(1), OptionalInt.of(1), Optional.<String>empty()},
+        new Object[] {
+          OptionalInt.of(2),
+          OptionalInt.of(1),
+          Optional.of("Minimum length must be less than or equal to the maximum length")
+        });
+  }
+
+  @Test
+  @Parameters(method = "getValidationTestData")
```

Comment:
**praise:** I love how readable this test class is.  Thanks for using the test parameters and avoiding duplication.

Reply:
As usual this was just copied from somewhere else, but thank you :)